File: 28-FindTheIndexOfTheFirstOccurrenceInAString/28-FindTheIndexOfTheFirstOccurrenceInAString.js
// Last updated: 4/29/2026, 10:36:21 AM
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function(haystack, needle) {
    let i = 0;
    let firstCh = needle[0];
    let firstOccurrence = -1;
    while(i < haystack.length) {
        let ch1 = haystack[i];
        if(ch1 === firstCh) {
            const start = i;
            let match = true;
            for(let j = 1; j < needle.length; j++) {
                ch1 = haystack[start + j];
                const ch2 = needle[j];
                if(ch1 !== ch2) {
                    match = false;
                    break;
                }
            }
            if(match) {
                firstOccurrence = start;
                return firstOccurrence;
            }
        }
        i++;
    }
    return firstOccurrence;
};

File: 28-FindTheIndexOfTheFirstOccurrenceInAString/28-FindTheIndexOfTheFirstOccurrenceInAString.py
# Last updated: 4/29/2026, 10:36:26 AM
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        def compare(i):
            for j in range(len(needle)):
                if i+j >= len(haystack):
                    return False
                if haystack[i+j] != needle[j]:
                    return False
            return True
        
        n_h = len(haystack)
        n_n = len(needle)
        for i in range(n_h):
            if compare(i):
                return i
        return -1